README
NAME
    WWW::Mechanize::Cached - Cache response to be polite

VERSION
    version 1.37

SYNOPSIS
        use WWW::Mechanize::Cached;

        my $cacher = WWW::Mechanize::Cached->new;
        $cacher->get( $url );

        # or, with your own Cache object
        use CHI;
        use WWW::Mechanize::Cached;
    
        my $cache = CHI->new(
            driver   => 'File',
            root_dir => '/tmp/mech-example'
        );
    
        my $mech = WWW::Mechanize::Cached->new( cache => $cache );
        $mech->get("http://www.google.com");

DESCRIPTION
    Uses the Cache::Cache hierarchy to implement a caching Mech. This lets
    one perform repeated requests without hammering a server impolitely.

    Repository: <http://github.com/oalders/www-mechanize-cached/tree/master>

CONSTRUCTOR
  new
    Behaves like, and calls, WWW::Mechanize's "new" method. Any params,
    other than those explicitly listed here are passed directly to
    WWW::Mechanize's constructor.

    You may pass in a "cache => $cache_object" if you wish. The
    *$cache_object* must have "get()" and "set()" methods like the
    "Cache::Cache" family.

    The default Cache object is set up with the following params:

        my $cache_params = {
            default_expires_in => "1d", namespace => 'www-mechanize-cached',
        };
    
        $cache = Cache::FileCache->new( $cache_params );

    This should be fine if you only want to use a disk-based cache, you only
    want to cache results for 1 day and you're not in a shared hosting
    environment. If any of this presents a problem for you, you should pass
    in your own Cache object. These defaults will remain unchanged in order
    to maintain backwards compatibility.

    For example, you may want to try something like this:

        use WWW::Mechanize::Cached;
        use CHI;
    
        my $cache = CHI->new(
            driver   => 'File',
            root_dir => '/tmp/mech-example'
        );
    
        my $mech = WWW::Mechanize::Cached->new( cache => $cache );
        $mech->get("http://www.google.com");

METHODS
    Most methods are provided by WWW::Mechanize. See that module's
    documentation for details.

  cache( $cache_object )
    Requires an caching object which has a get() and a set() method. Using
    the CHI module to create your cache is the recommended way. See new()
    for examples.

  is_cached()
    Returns true if the current page is from the cache, or false if not. If
    it returns "undef", then you don't have any current request.

  positive_cache( 0|1 )
    As of v1.36 positive caching is enabled by default. Up to this point,
    this module had employed a negative cache, which means it cached 404
    responses, temporary redirects etc. In most cases, this is not what you
    want, so the default behaviour now better reflects this. You can revert
    to the negative cache quite easily:

        # cache everything (404s, all 300s etc)
        $mech->positive_cache( 0 );

  ref_in_cache_key( 0|1 )
    Allow the referring URL to be used when creating the cache key. This is
    off by default. In almost all cases, you will not want to enable this,
    but it is available to you for reasons of backwards compatibility and
    giving you enough rope to hang yourself.

    Previous to v1.36 the following was in the "BUGS AND LIMITATIONS"
    section:

        It may sometimes seem as if it's not caching something. And this may well
        be true. It uses the HTTP request, in string form, as the key to the cache
        entries, so any minor changes will result in a different key. This is most
        noticable when following links as L<WWW::Mechanize> adds a C<Referer>
        header.

    See RT #56757 for a detailed example of the bugs this functionality can
    trigger.

THANKS
    Iain Truskett for writing this in the first place.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc WWW::Mechanize::Cached

    You can also look for information at:

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/WWW-Mechanize-Cached>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/WWW-Mechanize-Cached>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Mechanize-Cached>

    *   Search CPAN

        <http://search.cpan.org/dist/WWW-Mechanize-Cached>

PAST AUTHORS
    Iain Truskett <spoon@cpan.org>

    Maintained from 2004 - July 2009 by Andy Lester <petdance@cpan.org>

SEE ALSO
    WWW::Mechanize.

AUTHOR
    Olaf Alders <olaf@wundercounter.com> (current maintainer)

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Iain Truskett and Andy Lester.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module WWW::Mechanize::Cached

1.44 2014-08-23 11:25:15 America/Toronto
    - Support working with CHI automatically if Cache::FileCache is not installed
    - GH #6 (Kent Fredric)

1.43 2013-07-26 23:04:40 America/Toronto
    - Reformatted Changes file as per CPAN::Changes::Spec (Neil Bowers)

1.42 2013-06-29
    - Allow caching w/ Content-Encoding == chunked + no size header (Kent
      Fredric)

1.41 2012-06-27
    - Account for content length when deciding whether to cache (Kent Fredric)

1.40 2010-10-18
    - Fixed file prefix in cache_key.t when run under Windows

1.39 2010-10-09
    - live tests now skip if we can't connect to the live URL
    - Storable now has a specific version prerequisite

1.38 2010-10-07
    - cache_key test should now be portable to Windows
    - Critic test removed as Dist::Zilla runs this test automatically
    - POD tests should now only be run by author

1.37 2010-08-30
    - Removed TestCache from Makefile.  Had been automatically added
      by Dist::Zilla

1.36 2010-08-30
    - Caching is now positive by default (was negative up to this point)
    - Referring URLs are now, by defaut, not used when creating cache keys.
      This reverses the behaviour of previous versions.  RT #56757
    - Now using Moose and Dist::Zilla

1.35 2009-11-20
    - Changed test URL, which was causing a small percentage of installs to
      fail
    - Added latest HTTP::Headers to prerequisites in order to prevent some
      failed installs

1.34 2009-11-20
    - Added first example script
    - Expanded documentation re: custom caches
    - Removed some outdated files from distro

1.33 2009-07-13
    - Now maintained by Olaf Alders.  Thanks to Andy Lester for offering
      to let me take this on.
    - Applied patch from RT http://rt.cpan.org/Public/Bug/Display.html?id=42693
      This fixes the "Can't store CODE items" errors
    - Now requires Test::Warn 0.11 (fixes errors in t/002-bad-custom-cache.t)
    - Moved development to GitHub:
      http://github.com/oalders/www-mechanize-cached/tree/master
    - Added Perl::Critic author tests

1.32 2004-04-11

    [ENHANCEMENTS]
    - Added the $mech->is_cached() method.

    - Verified that reusing an existing cache gives you the pages back.

    [INTERNAL]
    - Now requires WWW::Mechanize 1.00, because previous versions
      goofed up the HTTP headers on some requests.


1.30 2004-03-14

    - No new features.

    [FIXED]
    - Fixed stupid configuration problem in t/bad-cache-test.t.
      If you got 1.28 installed OK, you don't need 1.30.

1.28 2004-03-13

    [THINGS THAT WILL BREAK YOUR CODE]
    - The C<cache> parm to the constructor is no longer a set
      of parms to specify how the cache will be initialized.  It
      must now be a fully-initialized cache object, probably in
      the Cache::Cache hierarchy.

    - Existing caches will not work, because I changed the
      directory that they get written to.  It used to go into
      /tmp/FileCache/WWW::Mechanize::Cached, but now will go into
      /tmp/FileCache/www-mechanize-cached.  This is so the Windows
      folks can use the module, too.

      However, if you want to create your own cache object that
      writes to /tmp/FileCache/WWW::Mechanize::Cached, for
      compatibility, you can create it and pass it in.

    [ENHANCEMENTS]
    - You can now specify your own Cache::Cache object to pass
      into the constructor.

1.26 2004-02-29

    [FIXES]
    - Removed a duplicate $key.  Cleaned up some of the internal code,
      and made the variable names more explicit.

    - Added an autocheck=>1 to the t/cached.t test.  Also hits
      time.gov instead of Iain's now-gone website.

    - Removed the SIGNATURE, which I don't intend to maintain.

1.24 2004-01-18

    NO NEW FUNCTIONALITY.

    On December 29th, 2003, Iain Truskett, the original author of
    WWW::Mechanize::Cached passed away.  I've taken over maintenance
    of the module.  If you're interested in carrying on Iain's work,
    let me know.


1.23 2003-12-01

    - Initial version, released by Iain Truskett.



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Cache response to be polite",
  "dist_author" => [
    "Iain Truskett (original author)",
    "Andy Lester <petdance\@cpan.org> (2004 - July 2009)",
    "Olaf Alders <olaf\@wundercounter.com> (current maintainer)"
  ],
  "dist_name" => "WWW-Mechanize-Cached",
  "dist_version" => "1.44",
  "license" => "perl",
  "module_name" => "WWW::Mechanize::Cached",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "CHI" => 0,
    "Cache::FileCache" => 0,
    "Carp" => 0,
    "Data::Dump" => 0,
    "Moose" => 0,
    "Storable" => "2.21",
    "WWW::Mechanize" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "Find::Lib" => 0,
    "HTTP::Request" => 0,
    "Path::Class" => 0,
    "Test::More" => 0,
    "Test::Requires" => 0,
    "constant" => 0,
    "lib" => 0,
    "vars" => 0
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "Find::Lib" => 0,
  "HTTP::Request" => 0,
  "Module::Build" => "0.28",
  "Path::Class" => 0,
  "Test::More" => 0,
  "Test::Requires" => 0,
  "constant" => 0,
  "lib" => 0,
  "vars" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Cache response to be polite",
  "AUTHOR" => "Iain Truskett (original author), Andy Lester <petdance\@cpan.org> (2004 - July 2009), Olaf Alders <olaf\@wundercounter.com> (current maintainer)",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "WWW-Mechanize-Cached",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "WWW::Mechanize::Cached",
  "PREREQ_PM" => {
    "CHI" => 0,
    "Cache::FileCache" => 0,
    "Carp" => 0,
    "Data::Dump" => 0,
    "Moose" => 0,
    "Storable" => "2.21",
    "WWW::Mechanize" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "Find::Lib" => 0,
    "HTTP::Request" => 0,
    "Path::Class" => 0,
    "Test::More" => 0,
    "Test::Requires" => 0,
    "constant" => 0,
    "lib" => 0,
    "vars" => 0
  },
  "VERSION" => "1.44",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CHI" => 0,
  "Cache::FileCache" => 0,
  "Carp" => 0,
  "Data::Dump" => 0,
  "File::Spec" => 0,
  "Find::Lib" => 0,
  "HTTP::Request" => 0,
  "Module::Build" => "0.28",
  "Moose" => 0,
  "Path::Class" => 0,
  "Storable" => "2.21",
  "Test::More" => 0,
  "Test::Requires" => 0,
  "WWW::Mechanize" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/WWW/Mechanize/Cached.pm
use strict;
use warnings FATAL => 'all';

package WWW::Mechanize::Cached;
$WWW::Mechanize::Cached::VERSION = '1.44';
use 5.006;

use Moose;
extends 'WWW::Mechanize';

use Carp qw( carp croak );
use Data::Dump qw( dump );
use Storable qw( freeze thaw );

has 'cache'                         => ( is => 'rw', );
has 'is_cached'                     => ( is => 'rw', );
has 'positive_cache'                => ( is => 'rw', );
has 'ref_in_cache_key'              => ( is => 'rw', );
has 'cache_undef_content_length'    => ( is => 'rw', );
has 'cache_zero_content_length'     => ( is => 'rw', );
has 'cache_mismatch_content_length' => ( is => 'rw', );
has '_verbose_dwarn'                => ( is => 'rw', );

sub new {
    my $class     = shift;
    my %mech_args = @_;

    my $cache = delete $mech_args{cache};
    if ( $cache ) {
        my $ok
            = ( ref( $cache ) ne "HASH" )
            && $cache->can( "get" )
            && $cache->can( "set" );
        if ( !$ok ) {
            carp "The cache param must be an initialized cache object";
            $cache = undef;
        }
    }

    my %cached_args = %mech_args;

    my %defaults = (
        ref_in_cache_key              => 0,
        positive_cache                => 1,
        cache_undef_content_length    => 0,
        cache_zero_content_length     => 0,
        cache_mismatch_content_length => 'warn',
        _verbose_dwarn                => 0,
    );

    for my $key ( keys %defaults ) {
        delete $mech_args{$key};
    }

    my $self = $class->SUPER::new( %mech_args );

    if ( !$cache ) {
        local $@;
        if ( eval { require Cache::FileCache; 1 } ) {
          my $cache_params = {
              default_expires_in => "1d",
              namespace          => 'www-mechanize-cached',
          };
          $cache = Cache::FileCache->new( $cache_params );
        } elsif ( eval { require CHI; 1 } ) {
          my $cache_params = {
            driver => 'File',
            expires_in => '1d',
            namespace => 'www-mechanize-cached',
          };
          $cache = CHI->new( %$cache_params );
        } else {
          croak("Could not create a default cache." .
            "Please make sure either CHI or Cache::FileCache are installed or configure manually as appropriate"
          );
        }
    }

    $self->cache( $cache );

    foreach my $arg ( keys %defaults ) {
        if ( exists $cached_args{$arg} ) {
            $self->$arg( $cached_args{$arg} );
        }
        else {
            $self->$arg( $defaults{$arg} );
        }
    }
    $self->is_cached( undef );

    return $self;
}

sub _make_request {

    my $self    = shift;
    my $request = shift;
    my $req     = $request;

    $self->is_cached( 0 );

    # An odd line to need.
    # No idea what purpose this serves?  OALDERS
    $self->{proxy} = {} unless defined $self->{proxy};

    # RT #56757
    if ( !$self->ref_in_cache_key ) {
        my $clone = $request->clone;
        $clone->header( Referer => undef );
        $req = $clone->as_string;
    }

    my $response = $self->cache->get( $req );

    if ( $response ) {
        $response = thaw( $response );
    }
    if ( $self->_cache_ok( $response ) ) {
        $self->is_cached( 1 );
        return $response;
    }

    $response = $self->SUPER::_make_request( $request, @_ );

    # decode strips some important headers.
    my $headers = $response->headers->clone;

    my $should_cache = $self->_response_cache_ok( $response, $headers );

    # http://rt.cpan.org/Public/Bug/Display.html?id=42693
    $response->decode();
    delete $response->{handlers};

    $self->cache->set( $req, freeze( $response ) ) if $should_cache;

    return $response;
}

sub _dwarn_filter {
    my ( $ctx, $ref ) = @_;
    return {
        hide_keys => [
            qw( _content cookie content set-cookie handlers cookie_jar cache req res page_stack )
        ]
    };

}

sub _dwarn {
    my $self    = shift;
    my $message = shift;

    return unless my $handler = $self->{onwarn};

    return if $self->quiet;

    if ( $self->_verbose_dwarn ) {
        my $payload = {
            self    => $self,
            message => $message,
            debug   => \@_,
        };
        require Data::Dump;
        return $handler->( Data::Dump::dumpf( $payload, \&_dwarn_filter ) );
    }
    else {
        return $handler->( $message );
    }
}

sub _response_cache_ok {
    my $self     = shift;
    my $response = shift;
    my $headers  = shift;

    return 0 if !$response;
    return 1 if !$self->positive_cache;

    return 0 if $response->code < 200;
    return 0 if $response->code > 301;

    if ( exists $headers->{'client-transfer-encoding'} ) {
        for my $cte ( @{ $headers->{'client-transfer-encoding'} } ) {

            # Transfer-Encoding = chunked means document consistency
            # is independent of Content-Length value,
            # and that Content-Length can be safely ignored.
            # Its not obvious how the lower levels represent a
            # failed chuncked-transfer yet.
            # But its safe to say relying on content-length proves pointless.
            return 1 if $cte eq 'chunked';
        }
    }

    my $size = $headers->{'content-length'};

    if ( not defined $size ) {
        if ( $self->cache_undef_content_length . q{} eq q{warn} ) {
            $self->_dwarn(
                q[Content-Length header was undefined, not caching]
                    . q[ (E=WWW_MECH_CACHED_CONTENTLENGTH_MISSING)],
                $headers
            );
            return 0;
        }
        if ( $self->cache_undef_content_length == 0 ) {
            return 0;
        }
    }

    if ( defined $size and $size == 0 ) {
        if ( $self->cache_zero_content_length . q{} eq q{warn} ) {
            $self->_dwarn(
                q{Content-Length header was 0, not caching}
                    . q{ (E=WWW_MECH_CACHED_CONTENTLENGTH_ZERO)},
                $headers
            );
            return 0;
        }
        if ( $self->cache_zero_content_length == 0 ) {
            return 0;
        }
    }

    if (    defined $size
        and $size != 0
        and $size != length( $response->content ) )
    {
        if ( $self->cache_mismatch_content_length . "" eq "warn" ) {
            $self->_dwarn(
                q{Content-Length header did not match contents actual length, not caching}
                    . q{ (E=WWW_MECH_CACHED_CONTENTLENGTH_MISSMATCH)} );
            return 0;
        }
        if ( $self->cache_mismatch_content_length == 0 ) {
            return 0;
        }
    }

    return 1;
}

sub _cache_ok {

    my $self     = shift;
    my $response = shift;

    return 0 if !$response;
    return 1 if !$self->positive_cache;

    return 0 if $response->code < 200;
    return 0 if $response->code > 301;

    return 1;
}

__PACKAGE__->meta->make_immutable( inline_constructor => 0 );

"We miss you, Spoon";    ## no critic

# ABSTRACT: Cache response to be polite

__END__

=pod

=encoding UTF-8

=head1 NAME

WWW::Mechanize::Cached - Cache response to be polite

=head1 VERSION

version 1.44

=head1 SYNOPSIS

    use WWW::Mechanize::Cached;

    my $cacher = WWW::Mechanize::Cached->new;
    $cacher->get( $url );

    # or, with your own Cache object
    use CHI;
    use WWW::Mechanize::Cached;

    my $cache = CHI->new(
        driver   => 'File',
        root_dir => '/tmp/mech-example'
    );

    my $mech = WWW::Mechanize::Cached->new( cache => $cache );
    $mech->get("http://www.wikipedia.org");

=head1 DESCRIPTION

Uses the L<Cache::Cache> hierarchy by default to implement a caching Mech. This
lets one perform repeated requests without hammering a server impolitely.
Please note that L<Cache::Cache> has been superceded by L<CHI>, but the default
has not been changed here for reasons of backwards compatibility.  For this
reason, you are encouraged to provide your own L<CHI> caching object to
override the default.

=head1 CONSTRUCTOR

=head2 new

Behaves like, and calls, L<WWW::Mechanize>'s C<new> method. Any params, other
than those explicitly listed here are passed directly to WWW::Mechanize's
constructor.

You may pass in a C<< cache => $cache_object >> if you wish. The
I<$cache_object> must have C<get()> and C<set()> methods like the
C<Cache::Cache> family.

The default Cache object is set up with the following params:

    my $cache_params = {
        default_expires_in => "1d", namespace => 'www-mechanize-cached',
    };

    $cache = Cache::FileCache->new( $cache_params );

This should be fine if you only want to use a disk-based cache, you only want
to cache results for 1 day and you're not in a shared hosting environment.
If any of this presents a problem for you, you should pass in your own Cache
object.  These defaults will remain unchanged in order to maintain backwards
compatibility.

For example, you may want to try something like this:

    use WWW::Mechanize::Cached;
    use CHI;

    my $cache = CHI->new(
        driver   => 'File',
        root_dir => '/tmp/mech-example'
    );

    my $mech = WWW::Mechanize::Cached->new( cache => $cache );
    $mech->get("http://www.wikipedia.org");

=head1 METHODS

Most methods are provided by L<WWW::Mechanize>. See that module's
documentation for details.

=head2 cache( $cache_object )

Requires an caching object which has a get() and a set() method. Using the CHI
module to create your cache is the recommended way. See new() for examples.

=head2 is_cached()

Returns true if the current page is from the cache, or false if not. If it
returns C<undef>, then you don't have any current request.

=head2 positive_cache( 0|1 )

As of v1.36 positive caching is enabled by default. Up to this point,
this module had employed a negative cache, which means it cached 404
responses, temporary redirects etc. In most cases, this is not what you want,
so the default behaviour now better reflects this. You can revert to the
negative cache quite easily:

    # cache everything (404s, all 300s etc)
    $mech->positive_cache( 0 );

=head2 ref_in_cache_key( 0|1 )

Allow the referring URL to be used when creating the cache key.  This is off
by default.  In almost all cases, you will not want to enable this, but it is
available to you for reasons of backwards compatibility and giving you enough
rope to hang yourself.

Previous to v1.36 the following was in the "BUGS AND LIMITATIONS" section:

    It may sometimes seem as if it's not caching something. And this may well
    be true. It uses the HTTP request, in string form, as the key to the cache
    entries, so any minor changes will result in a different key. This is most
    noticable when following links as L<WWW::Mechanize> adds a C<Referer>
    header.

See RT #56757 for a detailed example of the bugs this functionality can
trigger.

=head2 cache_undef_content_length( 0 | 'warn' | 1 )

This is configuration option which adjusts how caching behaviour performs when
the Content-Length header is not specified by the server.

Default behaviour is 0, which is not to cache.

Setting this value to 1, will cache pages even if the Content-Length header is
missing, which was the default behaviour prior to the addition of this feature.

And thirdly, you can set the value to the string 'warn', to warn if this
scenario occurs, and then not cache it.

=head2 cache_zero_content_length( 0 | 'warn' | 1 )

This is configuration option which adjusts how caching behaviour performs when
the Content-Length header is equal to 0.

Default behaviour is 0, which is not to cache.

Setting this value to 1, will cache pages even if the Content-Length header is
0, which was the default behaviour prior to the addition of this feature.

And thirdly, you can set the value to the string 'warn', to warn if this
scenario occurs, and then not cache it.

=head2 cache_mismatch_content_length( 0 | 'warn' | 1 )

This is configuration option which adjusts how caching behaviour performs when
the Content-Length header differs from the length of the content itself. ( Which
usually indicates a transmission error )

Setting this value to 0, will silenly not cache pages with a Content-Length
mismatch.

Setting this value to 1, will cache pages even if the Content-Length header
conflicts with the content length, which was the default behaviour prior to the
addition of this feature.

And thirdly, you can set the value to the string 'warn', to warn if this
scenario occurs, and then not cache it. ( This is the default behaviour )

=head1 UPGRADING FROM 1.40 OR EARLIER

Caching behaviour has changed since 1.40, and this may result in pages that were
previously cached start failing to cache, and in some cases, emit warnings.

To return to the 1.40 behaviour:

    $mech->cache_undef_content_length(1);  # Default is 0
    $mech->cache_zero_content_length(1);   # Default is 0
    $mech->cache_mismatch_content_length(1); # Default is 'warn'

=head1 THANKS

Iain Truskett for writing this in the first place.
Andy Lester for graciously handing over maintainership.
Kent Fredric for adding content length handling.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Mechanize::Cached

=over

=item * Search CPAN

L<https://metacpan.org/module/WWW::Mechanize::Cached>

=back

=head1 SEE ALSO

L<WWW::Mechanize>, L<WWW::Mechanize::Cached::GZip>.

=head1 AUTHORS

=over 4

=item *

Iain Truskett (original author)

=item *

Andy Lester <petdance@cpan.org> (2004 - July 2009)

=item *

Olaf Alders <olaf@wundercounter.com> (current maintainer)

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Iain Truskett and Andy Lester.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
